### ISSUE TITLE
Best way to handle imgData when cbinding

### ISSUE THREAD

**Issue by drighelli** [drighelli/SpatialExperiment#10]
I'm writing the `cbind` method, but I'm confused on how the `int_metadata` are cbinded into the SingleCellExperiment class producing multiple lists, one for each object passed as argument to the `cbind`.
Indeed the new int_metadata comes from  `int_m <- do.call(c, unname(lapply(args, int_metadata)))`.

The best to handle `imgData` into the `SpatialExperiment::cbind` is to have them rbinded them into one single `imgData`, but where to store it? Because we have a list of `int_metadata` elements.

Can't we store the `imgData` as a slot?

**Comment by LTLA** [comment]
**tl;dr** Yes, having a separate `imgData` slot is probably simplest. I had forgot that you would want to combine these.

There are much more exotic solutions involving structures similar to `S4Vectors::Factor`s. These have the advantage that the per-cell sample assignments are held in the same data structure as the sample metadata, which should avoid the need to coordinate sample identities across different parts of the `SpatialExperiment`. I was thinking of elaborating on these possibilities in this issue, but on further reflection, they are probably too complicated for users to deal with in `spe$samples`.

The other approach is to introduce a concept of samples in the SCE object itself, involving a 1:many mapping between samples and columns of the SCE. Then subclasses like `SpatialExperiment` could just store their additional image metadata there. However, this does start to run pretty close to the MAE's `sampleMap` concept - and honestly, there isn't much use for sample-level metadata in SCE itself, given that it doesn't cost much to expand it to per-cell metadata (which is easier to work with).

Just make sure you `unique()` the DF upon cbinding and you subset the DF upon subsetting.

**Comment by drighelli** [comment]
Thanks @LTLA for the answer.

About the sample identifiers storing, I was thinking about something similar because, as it is now, we replicate the sample_id in multiple sections of the object (into `colData(spe)$sample_id` and `imgData(spe)$sample_id`).
This is not so complicated to handle, exception made for the redundancy of the information and some additional code to give unicity of `sample_id` during the `cbind`.

A very simple solution would be to just store the `sample_id` as a `slot` and then point to it with a `pointer` (as in C), but I really don't know how to implement such a thing in R.

I think this would simplify a lot the `sample_id` unicity during operations such as `cbind` or `subset`.

**Comment by LTLA** [comment]
The `sample_id`s (and in fact, all character vectors) are _already_ pointers; this is how it is all implemented in R's string cache. In any case, if you were to use pointers, you'll have the equivalent problem of keeping all the pointers valid upon DataFrame reallocations during bind operations. Best to stick to character vectors and contract/expand the DF as necessary.

To be more precise about the contraction/expansion, this is desirable as we want to do something like:

```r
spe2 <- cbind(spe[,1:10], spe[,-(1:10)])
identical(nrow(imgData(spe)), nrow(imgData(spe2)))
```

In other words, avoid duplicates in the bind operation. Normally, this would be possible with just `unique(spe@imgData)`. But the grobs will mess that up, so it is probably necessary to remove the grob column before attempting this. This would imply that all other columns are sufficient to define the uniqueness of each `imgData` row; this should be contractually specified.

As a result of the above, it may be necessary to support arbitrary columns in the `imgData` DF if there are image attributes beyond resolution and whatever else we have mandated as floating around in there. For example, you could imagine different stainings, at which point you might need a `staining` column in the `imgData`. In fact, `scale_factor` could be considered an optional column that would only be relevant for Visium data; quite similar to `sizeFactors()` for SCE objects.

Also, why not `imageData`? Way more understandable, there's no law about three character prefixes in front of the `Data`.

**Comment by drighelli** [comment]
I understand your point, but this line is not always true
```
spe2 <- cbind(spe[,1:10], spe[,-(1:10)])
identical(nrow(imgData(spe)), nrow(imgData(spe2)))
```
You can have the case where multiple images are loaded in one `spe`.
Anyway, this is not a big problem, I handled the unicity by setting the sample_id(s) properly into the `cbind` (not a wonderful code, but it does what we need).

About the `scaleFactor`(s) and other `spatialCoordinates`, I avoided the problem by setting up a dynamic slot into the class which stores the names of the `colData` coordinates during the construction of the object.
In such a way we can handle the optional columns into the SPE.
Also in this case, this is not a wonderful approach, but it can be easily improved by saving these names into the `int_metadata` list or wherever.

**Comment by LTLA** [comment]
It _should_ always be true, even for multiple images. Note that the `unique()` considers all columns.

There is no need to have more slots, you just need to avoid constraining the columns of the `imageData`.

**Comment by drighelli** [comment]
Sorry, I still don't understand why the row number of the imgData should be identical between a singular object and a combined one.

**Comment by LTLA** [comment]
In my code above, I split an object and combine the two parts back together. The images should not be duplicated in the combined object, because they are the same images; hence the suggestion that you should be doing a `unique()` in the `cbind`, and the requirement for an explicit contract that the uniqueness of rows in the `imageData` is determined by all non-grob columns.
